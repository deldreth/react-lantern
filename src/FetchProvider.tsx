import React from 'react';

export interface Context {
  cache: Map<string, any>;
  baseUrl: string;
  operations: Map<string, any>;
  request: RequestInit;
  readonly updateRequest: (request: RequestInit) => void;
}

export const FetchContext = React.createContext<Context | null>(null);

export interface FetchProviderProps {
  baseUrl: string;
  children: React.ReactNode;
  defaultRequest?: RequestInit;
}

export default class FetchProvider extends React.Component<FetchProviderProps> {
  store: Context = {
    cache: new Map(),
    baseUrl: null,
    operations: new Map(),
    request: {},
    updateRequest: this.updateRequest.bind(this),
  };

  constructor(props: FetchProviderProps) {
    super(props);

    this.store.baseUrl = props.baseUrl;
    this.store.request = props.defaultRequest;
  }

  render() {
    return (
      <FetchContext.Provider value={this.store}>
        {this.props.children}
      </FetchContext.Provider>
    );
  }

  private updateRequest(request: RequestInit) {
    this.store.request = {
      ...this.store.request,
      ...request,
    };
  }
}
