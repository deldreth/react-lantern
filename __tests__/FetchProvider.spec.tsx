import React from 'react';

import { shallow } from 'enzyme';

import FetchProvider from '../src/FetchProvider';

test('FetchProvider baseUrl', () => {
  const Provider = shallow(
    <FetchProvider baseUrl="http://www.fireflyxd.com">
      <div>Test</div>
    </FetchProvider>
  );

  expect(Provider.props().value.baseUrl).toBe('http://www.fireflyxd.com');
  expect(Provider).toMatchSnapshot();
});
