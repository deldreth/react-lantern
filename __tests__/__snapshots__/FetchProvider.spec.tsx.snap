// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchProvider baseUrl 1`] = `
<ContextProvider
  value={
    Object {
      "baseUrl": "http://www.fireflyxd.com",
      "cache": Map {},
      "operations": Map {},
      "request": undefined,
      "updateRequest": [Function],
    }
  }
>
  <div>
    Test
  </div>
</ContextProvider>
`;
